packages/gguf/scripts/generate-llm.ts

```diff
@@ -0,0 +1,194 @@
+
+type MOE<TArchitecture extends LLMArchitecture> =
+	& { [K in \`\${TArchitecture}.expert_count\`]: number }
+	& { [K in \`\${TArchitecture}.expert_used_count\`]: number };
+
+export type TransformerLLMArchitecture = LLMArchitecture; // type alias
+export type TransformerLLMBase<TArchitecture extends LLMArchitecture> = ModelBase<TArchitecture>
+	& MOE<TArchitecture>
+	& Attention<TArchitecture>
+	& Rope<TArchitecture>;
+`;
+
+const KV_TYPE = {
+	LLM_KV_ATTENTION_LAYERNORM_RMS_EPS: "number",
+	LLM_KV_ATTENTION_LAYERNORM_EPS: "number",
+	LLM_KV_ATTENTION_CAUSAL: "boolean",
+	LLM_KV_TOKENIZER_TOKEN_TYPE_COUNT: "number",
+	LLM_KV_POOLING_TYPE: "number", // TODO: this is an enum
+	LLM_KV_ATTENTION_CLAMP_KQV: "number",
+	LLM_KV_ATTENTION_MAX_ALIBI_BIAS: "number",
+	LLM_KV_SSM_CONV_KERNEL: "number",
+	LLM_KV_SSM_INNER_SIZE: "number",
+	LLM_KV_SSM_STATE_SIZE: "number",
+	LLM_KV_SSM_TIME_STEP_RANK: "number",
+	LLM_KV_LOGIT_SCALE: "number",
+};
+
+interface Arch {
+	cppConst: string; // for example: "LLM_ARCH_LLAMA"
+	name: string; // for example: "llama"
+	tsName: string; // for example: "ArchLlama"
+	tensorNames: string[]; // for example: "token_embd"
+	hparams: string[];
+}
+
+async function main() {
+	const res = await fetch(SOURCE_CPP_URL);
+	const cppSource = await res.text();
+
+	/////////////////////////////////////
+	// extract list of all architectures
+	const archList: Arch[] = [];
+	const RE_ARCH_NAME = /LLM_ARCH_[A-Z0-9_]+/;
+	const matchedArchList = cppSource.match(/LLM_ARCH_NAMES = (?<names>[^;]+)/)?.groups?.names.split("\n") || [];
+	if (matchedArchList.length === 0) {
+		throw new Error("LLM_ARCH_NAMES is empty");
+	}
+	for (const line of matchedArchList) {
+		const matched = line.match(/(?<cppConst>LLM_ARCH_[A-Z0-9_]+),\s+"(?<name>.+?)"/);
+		if (matched && !matched.groups?.name.match(/unknown/)) {
+			archList.push({
+				cppConst: matched.groups?.cppConst || "",
+				name: matched.groups?.name || "",
+				tsName: snakeToPascal(matched[1].replace("LLM_", "")),
+				tensorNames: [],
+				hparams: [],
+			});
+		}
+	}
+
+	/////////////////////////////////////
+	// extract map constant name to kv name
+	// for example: LLM_KV_ATTENTION_LAYERNORM_RMS_EPS ==> "%s.attention.layer_norm_rms_epsilon"
+	const constToKVName: { [cppConst: string]: string } = {};
+	const matchedKVList = cppSource.match(/LLM_KV_NAMES = (?<names>[^;]+)/)?.groups?.names.split("\n") || [];
+	if (matchedKVList.length === 0) {
+		throw new Error("LLM_KV_NAMES is empty");
+	}
+	for (const line of matchedKVList) {
+		const matched = line.match(/(LLM_KV_[A-Z0-9_]+)[,\s]+"([^"]+)"/);
+		if (matched) {
+			constToKVName[matched[1]] = matched[2];
+		}
+	}
+
+	/////////////////////////////////////
+	// extract list of tensor names based on architecture
+	// TODO: unused for now
+	const matchedTensorList = cppSource.match(/LLM_TENSOR_NAMES = (?<names>[^;]+)/)?.groups?.names.split("\n") || [];
+	if (matchedTensorList.length === 0) {
```

Comment:
similar change is needed as in https://github.com/huggingface/huggingface.js/pull/640#discussion_r1583785883

Reply:
handled in 6962b5143a70631f5a2736b53ba1ddbdc8d609f0